[BUG] Niftisphere masker cannot be "refitted" once fitted

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Operating system

- [x] Linux
- [ ] Mac
- [ ] Windows

### Operating system version

For example one of the following:
- Linux Ubuntu 22.04


### Python version

- [ ] 3.13
- [x] 3.12
- [ ] 3.11
- [ ] 3.10
- [ ] 3.9

### nilearn version

dev

### Expected behavior

I am fairly sure most masker instances can be updated by refitting them, either by changing one of their attributes and refitting or by passing some different image to fit and refitting.

Niftisphere masker seems to be different as it will always return the first fitted instance.

I believe this is caused by this: https://github.com/nilearn/nilearn/blob/6fc1e42576a4c5ccc6542171bca419f64b21c25b/nilearn/maskers/nifti_spheres_masker.py#L522

### Current behavior & error messages

This is what I got:

```bash
AssertionError: masker.seeds_=[[0, 0, 0]]
```


### Steps and code to reproduce bug

```python
from nilearn.maskers import NiftiSpheresMasker

masker = NiftiSpheresMasker(seeds=[(0,0,0)])
masker.fit()
assert len(masker.seeds_) == 1

masker.seeds = [(0,0,0), (1,1,1)]
masker.fit()
assert len(masker.seeds_) == 2, f"{masker.seeds_=}"
```

--- a/nilearn/_utils/estimator_checks.py
+++ b/nilearn/_utils/estimator_checks.py
@@ -31,13 +31,17 @@ from nilearn.conftest import (
     _img_4d_rand_eye,
     _img_4d_rand_eye_medium,
     _img_4d_zeros,
+    _img_mask_eye,
     _make_surface_img,
+    _make_surface_mask,
     _rng,
     _shape_3d_default,
 )
 from nilearn.maskers import NiftiSpheresMasker
 from nilearn.regions import RegionExtractor
 from nilearn.reporting.tests.test_html_report import _check_html
+from nilearn.surface import SurfaceImage
+from nilearn.surface._testing import assert_surface_image_equal
 
 # List of sklearn estimators checks that are valid
 # for all nilearn estimators.
@@ -234,6 +238,7 @@ def nilearn_check_estimator(estimator):
         yield (clone(estimator), check_masker_clean_kwargs)
         yield (clone(estimator), check_masker_generate_report)
         yield (clone(estimator), check_masker_generate_report_false)
+        yield (clone(estimator), check_masker_refit)
 
         if not is_multimasker(estimator):
             yield (clone(estimator), check_masker_detrending)
@@ -408,6 +413,36 @@ def check_masker_clean(estimator):
     detrended_signal = estimator.fit_transform(input_img)
 
     assert_raises(AssertionError, assert_array_equal, detrended_signal, signal)
+
+
+def check_masker_refit(estimator):
+    """Check masker can be refitted and give different results."""
+    if accept_niimg_input(estimator):
+        mask_img_1 = _img_mask_eye()
+        mask_img_2 = _img_3d_ones()
+    else:
+        mask_img_1 = _make_surface_mask()
+        data = {}
+        for part in mask_img_1.data.parts:
+            data[part] = np.ones(mask_img_1.data.parts[part].shape)
+        mask_img_2 = SurfaceImage(mask_img_1.mesh, data)
+
+    estimator.mask_img = mask_img_1
+    estimator.fit()
+    fitted_mask_1 = estimator.mask_img_
+
+    estimator.mask_img = mask_img_2
+    estimator.fit()
+    fitted_mask_2 = estimator.mask_img_
+
+    if accept_niimg_input(estimator):
+        with pytest.raises(AssertionError):
+            assert_array_equal(
+                fitted_mask_1.get_fdata(), fitted_mask_2.get_fdata()
+            )
+    else:
+        with pytest.raises(AssertionError):
+            assert_surface_image_equal(fitted_mask_1, fitted_mask_2)
 
 
 # ------------------ SURFACE MASKER CHECKS ------------------

--- a/nilearn/_utils/estimator_checks.py
+++ b/nilearn/_utils/estimator_checks.py
@@ -31,7 +31,6 @@ from nilearn.conftest import (
     _img_4d_rand_eye,
     _img_4d_rand_eye_medium,
     _img_4d_zeros,
-    _img_mask_eye,
     _make_surface_img,
     _make_surface_mask,
     _rng,
@@ -418,8 +417,17 @@ def check_masker_clean(estimator):
 def check_masker_refit(estimator):
     """Check masker can be refitted and give different results."""
     if accept_niimg_input(estimator):
-        mask_img_1 = _img_mask_eye()
-        mask_img_2 = _img_3d_ones()
+        # using larger images to be compatible
+        # with regions extraction tests
+        # TODO refactor a common fixture for "large 3D shape"
+        shape = (29, 30, 31)
+        mask = np.zeros(shape, dtype=np.int8)
+        mask[1:-1, 1:-1, 1:-1] = 1
+        mask_img_1 = Nifti1Image(mask, _affine_eye())
+
+        mask = np.zeros(shape, dtype=np.int8)
+        mask[3:-3, 3:-3, 3:-3] = 1
+        mask_img_2 = Nifti1Image(mask, _affine_eye())
     else:
         mask_img_1 = _make_surface_mask()
         data = {}

--- a/nilearn/maskers/nifti_spheres_masker.py
+++ b/nilearn/maskers/nifti_spheres_masker.py
@@ -528,9 +528,6 @@ class NiftiSpheresMasker(BaseMasker):
         All parameters are unused; they are for scikit-learn compatibility.
 
         """
-        if hasattr(self, "seeds_"):
-            return self
-
         self._report_content = {
             "description": (
                 "This reports shows the regions defined "

--- a/nilearn/maskers/surface_maps_masker.py
+++ b/nilearn/maskers/surface_maps_masker.py
@@ -179,16 +179,22 @@ class SurfaceMapsMasker(_BaseSurfaceMasker):
                 verbose=self.verbose,
             )
             check_same_n_vertices(self.maps_img.mesh, self.mask_img.mesh)
+            self.mask_img_ = self.mask_img
             # squeeze the mask data if it is 2D and has a single column
-            for part in self.mask_img.data.parts:
+            for part in self.mask_img_.data.parts:
                 if (
-                    self.mask_img.data.parts[part].ndim == 2
-                    and self.mask_img.data.parts[part].shape[1] == 1
+                    self.mask_img_.data.parts[part].ndim == 2
+                    and self.mask_img_.data.parts[part].shape[1] == 1
                 ):
-                    self.mask_img.data.parts[part] = np.squeeze(
-                        self.mask_img.data.parts[part], axis=1
+                    self.mask_img_.data.parts[part] = np.squeeze(
+                        self.mask_img_.data.parts[part], axis=1
                     )
-            self.mask_img.data._check_ndims(1, "mask_img")
+            self.mask_img_.data._check_ndims(1, "mask_img")
+        else:
+            # TODO
+            # self.mask_img_ should be a SurfaceImage instance
+            # after fit
+            self.mask_img_ = None
 
         self._shelving = False
 
@@ -216,8 +222,8 @@ class SurfaceMapsMasker(_BaseSurfaceMasker):
             )
 
         self._reporting_data = {
-            "maps_img": self.maps_img,
-            "mask": self.mask_img,
+            "maps_img": self.maps_img_,
+            "mask": self.mask_img_,
             "images": None,  # we will update image in transform
         }
 

